brainiak/matnormal/covs.py

```diff
@@ -427,43 +435,57 @@ class CovUnconstrainedInvCholesky(CovBase):
        this saves a cholesky solve on every step of optimization
     """
 
-    def __init__(self, size, invSigma=None):
+    def __init__(self, size=None, invSigma=None):
+
+        if size is None and invSigma is None:
+            raise RuntimeError(
+                "Must pass either invSigma or size but not both")
+
+        if size is not None and invSigma is not None:
+            raise RuntimeError(
+                "Must pass either invSigma or size but not both")
+
+        if invSigma is not None:
+            size = invSigma.shape[0]
+
         super(CovUnconstrainedInvCholesky, self).__init__(size)
 
+        # number of parameters in the triangular mat
+        npar = (size * (size + 1)) // 2
+
         if invSigma is None:
-            self.Linv_full = tf.Variable(
-                tf.random.normal([size, size], dtype=tf.float64),
-                name="Linv_full")
+            self.Linv_flat = tf.Variable(
+                tf.random.normal([npar], dtype=tf.float64), name="Linv_flat"
+            )
+
         else:
-            self.Linv_full = tf.Variable(
-                np.linalg.cholesky(invSigma), name="Linv_full")
-
-        # Zero out triu of L_full to get cholesky L.
-        # This seems dumb but TF is smart enough to set the gradient to zero
-        # for those elements, and the alternative (fill_lower_triangular from
-        # contrib.distributions) is inefficient and recommends not doing the
-        # packing (for now).
-        # Also: to make the parameterization unique we log the diagonal so
-        # it's positive.
-        L_indeterminate = tf.linalg.band_part(self.Linv_full, -1, 0)
-        self.Linv = tf.linalg.set_diag(
-            L_indeterminate, tf.exp(tf.linalg.diag_part(L_indeterminate))
-        )
+            Linv = np.linalg.cholesky(invSigma)
+            self.Linv_flat = tf.Variable(
+                flatten_cholesky_unique(Linv), name="Linv_flat"
+            )
+
+    @property
+    def Linv(self):
+        """
+        Inverse of Cholesky factor of this covariance
+        """
+        return unflatten_cholesky_unique(self.Linv_flat)
 
     @property
     def logdet(self):
-        return -2 * tf.reduce_sum(input_tensor=tf.math.log(tf.linalg.diag_part(self.Linv)))
+        return -2 * tf.reduce_sum(
+            input_tensor=tf.math.log(tf.linalg.diag_part(self.Linv))
```

Comment:
Same concern about negative value in Linv.

Reply:
Same response -- also defined via unique cholesky parameterization.